src/main/java/betteradvancements/gui/GuiBetterAdvancement.java

```diff
@@ -52,21 +56,144 @@ public GuiBetterAdvancement(GuiBetterAdvancementTab guiBetterAdvancementTab, Min
             k = mc.fontRenderer.getStringWidth("  ") + mc.fontRenderer.getStringWidth("0") * strLengthRequirementCount * 2 + mc.fontRenderer.getStringWidth("/");
         }
         int titleWidth = 29 + mc.fontRenderer.getStringWidth(this.title) + k;
+        this.criterionGrid = this.getCriterionGrid();
+        int maxWidth = Math.max(titleWidth, this.criterionGrid.width);
         String s = displayInfo.getDescription().getFormattedText();
-        this.description = this.findOptimalLines(s, titleWidth);
+        this.description = this.findOptimalLines(s, maxWidth);
 
         for (String line : this.description) {
-            titleWidth = Math.max(titleWidth, mc.fontRenderer.getStringWidth(line));
+            maxWidth = Math.max(maxWidth, mc.fontRenderer.getStringWidth(line));
+        }
+
+        this.width = maxWidth + 8;
+    }
+
+    private CriterionGrid getCriterionGrid() {
+        if (advancementProgress == null || advancementProgress.isDone()) {
+            return new CriterionGrid();
+        }
+        Map<String, Criterion> criteria = advancement.getCriteria();
+        if (criteria.size() <= 1) {
+            return new CriterionGrid();
+        }
+        boolean anyObtained = false;
+        int numUnobtained = 0;
+        List<String> cellContents = new ArrayList<String>();
+        for (String criterion : criteria.keySet()) {
+            if (advancementProgress.getCriterionProgress(criterion).isObtained()) {
+                cellContents.add(" �2+�  " + criterion);
+                anyObtained = true;
+            }
+            else {
+                numUnobtained++;
+            }
+        }
+        if (!anyObtained) {
+            return new CriterionGrid();
+        }
+        cellContents.add(" �4x�  �o" + numUnobtained + " remaining");
+
+        GuiScreenBetterAdvancements screen = guiBetterAdvancementTab.getScreen();
+        int[] cellWidths = new int[cellContents.size()];
+        for (int i = 0; i < cellWidths.length; i++) {
+            cellWidths[i] = this.minecraft.fontRenderer.getStringWidth(cellContents.get(i));
+        }
+
+        double maxAspectRatio = (double)screen.width / screen.height;
+        CriterionGrid prevGrid = null;
+        for (int numCols = 1; numCols <= cellContents.size(); numCols++)
+        {
+            CriterionGrid currGrid = new CriterionGrid(cellContents, cellWidths, this.minecraft.fontRenderer.FONT_HEIGHT, numCols);
+            if (prevGrid != null && currGrid.numRows == prevGrid.numRows)
+                continue; // We increased the width without decreasing the height, which is pointless.
+            currGrid.init();
+            if (currGrid.aspectRatio > maxAspectRatio) {
+                if (prevGrid == null) {
+                    prevGrid = currGrid;
+                }
+                break;
+            }
+            prevGrid = currGrid;
+        }
+        return prevGrid;
+    }
+
+    private class CriterionGrid {
```

Comment:
Please move this inner class to the util package.

Reply:
Will do!